src/coreclr/src/vm/jitinterface.cpp

```diff
@@ -3848,7 +3820,7 @@ BOOL CEEInfo::canInlineTypeCheckWithObjectVTable (CORINFO_CLASS_HANDLE clsHnd)
 
     if (VMClsHnd.IsTypeDesc())
     {
-        // We can't do this optimization for arrays because of the object methodtable is template methodtable
+        // We can't do this optimization for TypeDesc because methodtable could be shared
```

Comment:
some TypeDescs still share method tables. I am not sure if such TypeDescs can occur on this code path, but perhaps it is safer to keep the check?

---
In reply to: [362353144](https://github.com/dotnet/runtime/pull/1201#discussion_r362353144) [](ancestors = 362353144)

Reply:
The JIT or JITed code will never see the MethodTable behind these TypeDescs. I think it should be safe to delete this.

Reply:
Will make `VMClsHnd.IsTypeDesc()` an assert

---
In reply to: [362378456](https://github.com/dotnet/runtime/pull/1201#discussion_r362378456) [](ancestors = 362378456)

Reply:
I do not think you can assert this. The JIT may pass TypeDesc into this method in some cases.

My point was the JIT or JITed code never sees the underlying MethodTable that may be shared.

Reply:
I see. It is possible to see a MethodDesc with shared MethodTable (ex: `typeof(int*) == o.GetType()`), but that is ok now. That is because optimized version compares literally _handle_ to the methodtable of the object and that produce valid result since none of remaining MethodDesc types can have heap instances.
We could as well just "optimize" the compare into `false` when the handle comparand is a TypeDesc, but it is such a corner case, so not worth the trouble.

---
In reply to: [362417850](https://github.com/dotnet/runtime/pull/1201#discussion_r362417850) [](ancestors = 362417850)